ldms/src/sampler/llnl/lustre_client.c

```diff
@@ -0,0 +1,641 @@
+
+        /* Make sure we have llite_data objects in the new_llite_tree for
+           each currently existing directory.  We can find the objects
+           cached in the global llite_tree (in which case we move them
+           from llite_tree to new_llite_tree), or they can be newly allocated
+           here. */
+
+        dir = opendir(LLITE_PATH);
+        if (dir == NULL) {
+                log_fn(LDMSD_LWARNING, SAMP" unable to open llite dir %s\n",
+                       LLITE_PATH);
+                return;
+        }
+        while ((dirent = readdir(dir)) != NULL) {
+                struct rbn *rbn;
+                struct llite_data *llite;
+
+                if (dirent->d_type != DT_DIR ||
+                    strcmp(dirent->d_name, ".") == 0 ||
+                    strcmp(dirent->d_name, "..") == 0)
+                        continue;
+                rbn = rbt_find(&llite_tree, dirent->d_name);
+                if (rbn) {
+                        llite = container_of(rbn, struct llite_data,
+                                           llite_tree_node);
+                        rbt_del(&llite_tree, &llite->llite_tree_node);
+                } else {
+                        llite = llite_create(dirent->d_name, LLITE_PATH);
+                }
+                if (llite == NULL)
+                        continue;
+                rbt_ins(&new_llite_tree, &llite->llite_tree_node);
+        }
+        closedir(dir);
+
+        /* destroy any llites remaining in the global llite_tree since we
+           did not see their associated directories this time around */
+        /* llites_destroy();*/
+        /* NOTE: ldms_set_delete() doesn't really work until some time post v4.2.
+           So instead of llites_destroy(), we'll just keep old data around forever
+           in this version of the code.  When ldms_set delete() works we can
+           delete the following while() loop and uncomment the llites_destroy()
+           call above. */
+        while (!rbt_empty(&llite_tree)) {
+                struct rbn *rbn;
+                rbn = rbt_min(&llite_tree);
+                rbt_del(&llite_tree, rbn);
+                rbt_ins(&new_llite_tree, rbn);
+        }
+
+        /* copy the new_llite_tree into place over the global llite_tree */
+        memcpy(&llite_tree, &new_llite_tree, sizeof(struct rbt));
+
+        return;
+}
+
+static void llites_sample()
+{
+        struct rbn *rbn;
+
+        /* walk tree of known LLITEs */
+        RBT_FOREACH(rbn, &llite_tree) {
+                struct llite_data *llite;
+                llite = container_of(rbn, struct llite_data, llite_tree_node);
+                llite_general_sample(llite->name, llite->stats_path,
+                                   llite->general_metric_set);
+        }
+}
+
+static int sample(struct ldmsd_sampler *self)
+{
+        log_fn(LDMSD_LDEBUG, SAMP" sample() called\n");
+        if (llite_general_schema_is_initialized() < 0) {
```

Comment:
this check can be omitted if config() is right.

Reply:
See other comment on this subject.